util/src/test-support/java/tech/pegasys/artemis/util/executors/MockExecutorService.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.executors;
+
+import static tech.pegasys.artemis.util.async.FutureUtil.ignoreFuture;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class MockExecutorService implements ExecutorService {
+
+  private boolean autoRun = false;
+
+  private final List<ExecutorTask<?>> tasks = new ArrayList<>();
+
+  // Test utility for inspecting executor's futures
+  public List<Future<?>> getFutures() {
+    return tasks.stream().map(ExecutorTask::getFuture).collect(Collectors.toList());
+  }
+
+  public void setAutoRun(final boolean shouldAutoRunTasks) {
+    this.autoRun = shouldAutoRunTasks;
+  }
+
+  public void runPendingFutures() {
+    final List<ExecutorTask<?>> currentTasks = new ArrayList<>(tasks);
```

Comment:
do you not just do `tasks.forEach()` to make sure to not run futures added after you called this method?

Reply:
Right - we only want to run tasks that are currently tracked.  Some of these tasks might trigger additional scheduled tasks.